src/main/java/com/google/firebase/internal/FirebaseExecutors.java

```diff
@@ -16,20 +16,122 @@
 
 package com.google.firebase.internal;
 
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.google.firebase.FirebaseApp;
+import com.google.firebase.ThreadManager;
+
+import java.util.HashSet;
+import java.util.Set;
 import java.util.concurrent.Executors;
 import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadFactory;
 
-/** Default executors used for internal Firebase threads. */
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Default ThreadManager implementations used by the Admin SDK. */
 public class FirebaseExecutors {
 
-  public static final ScheduledExecutorService DEFAULT_SCHEDULED_EXECUTOR;
+  private static final Logger logger = LoggerFactory.getLogger(FirebaseExecutors.class);
+
+  public static final ThreadManager DEFAULT_THREAD_MANAGER;
 
   static {
     if (GaeThreadFactory.isAvailable()) {
-      DEFAULT_SCHEDULED_EXECUTOR = GaeThreadFactory.DEFAULT_EXECUTOR;
+      DEFAULT_THREAD_MANAGER = new GaeThreadManager();
     } else {
-      DEFAULT_SCHEDULED_EXECUTOR =
-          Executors.newSingleThreadScheduledExecutor(Executors.defaultThreadFactory());
+      DEFAULT_THREAD_MANAGER = new DefaultThreadManager();
+    }
+  }
+
+  /**
+   * An abstract ThreadManager implementation that uses the same executor service
+   * across all active apps. The executor service is initialized when the first app is initialized,
+   * and terminated when the last app is deleted. This class is thread safe.
+   */
+  abstract static class GlobalThreadManager extends ThreadManager {
+
+    private final Set<String> apps = new HashSet<>();
+    private ScheduledExecutorService executorService;
+
+    @Override
+    protected synchronized ScheduledExecutorService getExecutor(FirebaseApp app) {
```

Comment:
I would suggest not to synchronize on the public facing object (since this can also be done by consumers). Can you use a private member (like executorService)?

Reply:
Done